packages/flutter/lib/src/material/floating_action_button.dart

```diff
@@ -186,4 +204,71 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
 
     return result;
   }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    Scaffold.setFloatingActionButtonNotchMakerFor(context, _notchMaker);
+  }
+
+  @override
+  void deactivate() {
+    Scaffold.setFloatingActionButtonNotchMakerFor(context, null);
+    super.deactivate();
+  }
+
+  Path _notchMaker(Rect host, Rect guest, Offset start, Offset end) {
+  if (end.dy != host.top)
```

Comment:
these checks should be in an assert block (or just make them asserts)

Reply:
same applies to all the exception-throwing code below.

Reply:
done